core/src/tile/tileManager.cpp

```diff
@@ -39,8 +40,18 @@ TileManager::~TileManager() {
     m_tileSets.clear();
 }
 
-void TileManager::addDataSource(std::shared_ptr<DataSource>&& dataSource) {
+int32_t TileManager::addDataSource(std::shared_ptr<DataSource>&& dataSource) {
     m_tileSets.push_back({++m_tileSetSerial, dataSource});
+    return m_tileSetSerial;
+}
+
+std::shared_ptr<ClientGeoJsonSource> TileManager::getClientSourceById(int32_t _id) {
+    for (auto& set : m_tileSets) {
+        if (set.id == _id) {
+            return std::dynamic_pointer_cast<ClientGeoJsonSource>(set.source);
```

Comment:
Evidently `dynamic_pointer_cast` creates a new `shared_ptr` too, so this is probably fine the way it is.